platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp

```diff
@@ -0,0 +1,379 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2023 Technology Innovation Institute.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file px4io_serial.cpp
+ *
+ * Serial interface for PX4IO on MPFS
+ */
+
+#include <syslog.h>
+
+#include <px4_arch/px4io_serial.h>
+#include <termios.h>
+#include <debug.h>
+
+/* Nb. this register only exists in FPGA UART implementation. For MSS uart
+ * this is a scratch register, and the code works by IRQ on each byte
+ */
+
+#define MPFS_UART_RFT_OFFSET 0x1C
+
+uint8_t ArchPX4IOSerial::_io_buffer_storage[sizeof(IOPacket)];
+
+ArchPX4IOSerial::ArchPX4IOSerial() :
+	_current_packet(nullptr),
+	_completion_semaphore(SEM_INITIALIZER(0))
+{
+}
+
+ArchPX4IOSerial::~ArchPX4IOSerial()
+{
+
```

Comment:
If you have interest you could remove these extra line feeds. Makes no difference to me. OCD might trigger for some.